md/toi13_robots.md

```diff
@@ -0,0 +1,100 @@
+สำหรับข้อนี้ โจทย์ต้องการให้หุ่นยนต์ไปเก็บของแต่ละชิ้น
+
+### Observation 1
+
+เราสามารถมองได้ว่า เราต้องการให้ของทุกชิ้น ได้รับการเก็บโดยหุ่นยนต์ 1 ตัว นั่นคือ เราไม่สนใจว่าหุ่นยนต์ตัวไหนจะมาเก็บของชิ้นปัจจุบัน แต่เราต้องการให้หุ่นยนต์ **ที่อยู่ใกล้ที่สุด** เป็นหุ่นยนต์เก็บของชิ้นนี้ 
+
+ตอนนี้โจทย์จะกลายเป็นว่า สำหรับของแต่ละชิ้น ให้หาหุ่นยนต์ที่ใกล้ที่สุด (ถ้ามี) แล้วนำระยะทางใกล้สุดของของแต่ละชิ้นมาบวกกัน แล้วนับด้วยว่ามีของอยู่กี่ชิ้นที่สามารถหาหุ่นยนต์ใกล้สุดได้ (กรณีที่หาไม่ได้เช่น มีกำแพงล้อมรอบ)
+
+วิธีการที่เห็นได้ชัดเจนที่สุด คือการทำ *Breadth-first search* จากของแต่ละชิ้นไปหาหุ่นยนต์ที่ใกล้ที่สุด ซึ่งจะใช้เวลา $\mathcal{O}(TNM)$ หากมีของทั้งหมด $T$ ชิ้น
+
+อีกวิธีที่มีความเป็นไปได้คือหาว่าหุ่นยนต์แต่ละตัวมีระยะทางไปแต่ละจุดเท่าใด แล้วค่อยๆตามมาดูทีหลังว่าของแต่ละชิ้นควรจะไปเชื่อมกับหุ่นยนต์ตัวไหน ใช้เวลา $\mathcal{O}(KNM + TK)$ หากมีหุ่นยนต์ $K$ ตัว
+
+วิธีการข้างต้นดูเหมือนเป็นวิธีการที่แย่กว่าวิธีแรก แต่จะทำให้ปรับความเข้าใจสู่ข้อสังเกตต่อไปได้โดยง่าย
+
+### Observation 2
+
+สมมติให้ $d(i, j)$ แทนระยะห่าง ระหว่างหุ่นยนต์ตัวที่ $i$ กับ ของชิ้นที่ $j$ สิ่งที่เราต้องการหา คือ $\sum\limits_{j=1}^{T}{\min\limits_{i=1}^{K}{d(i,j)}}$
+
+สังเกตว่า ไม่ว่า $i$ จะเป็นหุ่นยนต์ตัวไหนก็ตาม หากระยะทางสั้นสุด ก็ทำให้ได้คำตอบตามที่ต้องการ
+
+ปัญหาดังกล่าว คือ ปัญหา *Multiple Source Shortest Path (MSSP)* ซึ่งมีวิธีการแก้ปัญหาได้เร็วกว่าการทดลองเริ่มต้นจากแต่ละจุดที่ต่างกัน โดยจะเปลี่ยนมาเริ่มจาก **ทุกจุดพร้อมกัน** แทน
+
+การกระทำดังกล่าว อาจดูเหมือนเป็นไปไม่ได้ แต่เราสามารถดัดแปลงกระบวนการ *Breadth-first search* ได้ดังนี้
+
+ตัวอย่างโค้ดการ *Breadth-first search* บนตาราง แบบปกติ
+
+```cpp
+int dist[2048][2048]; // dist[i][j] := distance between the current robot and point (row i, col j)
```

Comment:
ทำไมเม้นเป็น eng อะ?

Reply:
Comments inside code blocks are in english to prevent (potential) file encoding errors in some compilers.

Reply:
เก่งมาก

Reply:
I don't think so. You should comment it in the Thai language.

Reply:
เหงาหรอครับ วดฟ

Reply:
npm, I will merge it.